### ISSUE TITLE
Strategy for generating binary output

### ISSUE THREAD

**Issue by jspc** [vinyl-linux/mint#3]
For some stuff this is easy; some fields have fixed lengths for instance (your basic primitive scalars; integers, chars, bools, whatevers). Other stuff, like collections (arrays, maps) less so.

To investigate:

1. Should each message contain a header of sizes?
2. Should collections be `nul` terminated?

A header might make smaller messages larger than they need to be, but larger messages slower (when we need to consume a stream byte-by-byte until we get a `nul`).

Of course, allowing sized collections (such as char array strings for things like UUIDs, where we know the size is fixed at 128bits) would work.

**Comment by jspc** [comment]
We can use `encoding/binary.Read` and `encoding/binary.Write` for anything with a fixed size, and iterate on an `io.Reader` for stuff like strings, looking for `null` values to terminate.

Maps and slices work much the same:

1. For a slice iterate through until we get a `null`
2. For a map iterate through two by two, until we get a `null`

**Comment by jspc** [comment]
1. For slices we need to know the length when `Marshalling`/`Unmarshalling`; fixed length slices are fine- we can generate this. For non-fixed, serialise the length as a `uint32` prior to the payload. This caps slice lengths to ~4 million or whatever, but we'll see
2. Maps are represented as an unfixed length slice of `[k0, v0, k1, v1....kn, vn]`

**Comment by jspc** [comment]
Unmarshalling a large string by reading byte-by-byte until we get a `null` is slow. Much slower than Marshalling, much slower than type asserting the result of `func` Value() any`.

Perhaps it's time to treat a string as a byte array, and using the same strategy we use above.

**Comment by jspc** [comment]
Unscientific numbers from:

1. Reading/ writing strings byte by byte with a null terminator; and
2. Turning strings into byte slices, prefixing with a size

## tl;dr

Nahhhh.... if Marshalling a 600MB string takes a few seconds and the OOMKiller doesn't get involved then fine. The same string with a byte array is many, many times worse then why bother changing?

## Byte by byte
<details><summary>Test output</summary>

```
--- PASS: TestStringScalar (8.62s)                                                                                                                                                            
    --- PASS: TestStringScalar/Massive_string_should_return_exactly (7.99s)
        --- PASS: TestStringScalar/Massive_string_should_return_exactly/Marshall (0.14s)
        --- PASS: TestStringScalar/Massive_string_should_return_exactly/Unmarshall (7.81s)
        --- PASS: TestStringScalar/Massive_string_should_return_exactly/Value (0.05s)
```
</details> 

Unmarshalling takes ~8 seconds, whereas other operations take fractions of a second.

## As a slice collection
<details><summary>Test output</summary>

```
--- PASS: TestStringScalar (84.91s)
    --- PASS: TestStringScalar/Massive_string_should_return_exactly (84.29s)
        --- PASS: TestStringScalar/Massive_string_should_return_exactly/Marshall (17.00s)
        --- PASS: TestStringScalar/Massive_string_should_return_exactly/Unmarshall (64.75s)
        --- PASS: TestStringScalar/Massive_string_should_return_exactly/Value (2.53s)
```

Unmarshalling takes over a minute, marshalling takes a cool 17 seconds. Even getting the `Value()` takes a couple of seconds.

Odd times, too, the oomkiller nukes the test runner

</details>

**Comment by jspc** [comment]
Fixed in #7
